### PR TITLE
Mat4 and transformations

### PR DESCRIPTION
**Pull request by LevSmirnov1997**
Very basic coding

### CODE REVIEW

File: src/mat4.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    values[0] = values[5] = values[10] = values[15] = x;
+}
+
+mat4::mat4(float arr[16]) {
+    std::copy(arr, arr + 16, values);
+}
+
+mat4 mult(const mat4& lhs, const mat4& rhs) {
+    float t_arr[16];
+    std::fill_n(t_arr, 16, 0);
```
Comment (DebTalen) [inline src/mat4.cpp]: use = { 0.0f } to initialize an array

File: src/mat_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#include <mat_utils.hpp>
+#include <algorithm>
+#include <cmath>
+
+mat4 scale(const mat4& transf, const vec2& sc) {
+    float t_arr[16];
+    std::fill_n(t_arr, 16, 0);
```
Comment (DebTalen) [inline src/mat_utils.cpp]: use = { 0.0f } to initialize an array

File: src/mat4.cpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#include "mat4.hpp"
+#include <algorithm>
+
+mat4::mat4() {
+    std::fill_n(values, 16, 0);
```
Comment (DebTalen) [inline src/mat4.cpp]: use values { 0.0f } in member initialization list

File: src/mat4.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+mat4::mat4(float x) {
+    std::fill_n(values, 16, 0);
```
Comment (DebTalen) [inline src/mat4.cpp]: use values { 0.0f } in member initialization list

File: src/mat_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    t_arr[0] = sc.get_x();
+    t_arr[5] = sc.get_y();
+    t_arr[10] = 1.0f;
+    t_arr[15] = 1.0f;
+    return (mult(mat4(t_arr), transf));
```
Comment (DebTalen) [inline src/mat_utils.cpp]: remove redundant parentheses

File: src/mat_utils.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+mat4 rotate(const mat4& transf, float rt) {
+    float t_arr[16];
+    std::fill_n(t_arr, 16, 0);
```
Comment (DebTalen) [inline src/mat_utils.cpp]: use { 0.0f } to initialize an array

File: src/mat_utils.cpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    for(int i = 0; i < 4; ++i)
+        t_arr[5*i] = 1;
+    t_arr[0] = cos(rt * 3.141592653 / 180);
```
Comment (DebTalen) [inline src/mat_utils.cpp]: define pi as a constexpr value

File: src/mat_utils.cpp [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    t_arr[1] = -sin(rt * 3.141592653 / 180);
+    t_arr[4] = sin(rt * 3.141592653 / 180);
+    t_arr[5] = cos(rt * 3.141592653 / 180);
+    return (mult(mat4(t_arr), transf));
```
Comment (DebTalen) [inline src/mat_utils.cpp]: remove redundant parentheses

File: src/mat_utils.cpp [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+mat4 translate(const mat4& transf, const vec2& tr) {
+    float t_arr[16];
+    std::fill_n(t_arr, 16, 0);
```
Comment (DebTalen) [inline src/mat_utils.cpp]: use { 0.0f } to initialize an array

File: src/mat_utils.cpp [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    for(int i = 0; i < 4; ++i)
+        t_arr[5*i] = 1;
+    t_arr[3] = tr.get_x();
+    t_arr[7] = tr.get_y();
+    t_arr[11] = 1.0f;
+    return (mult(mat4(t_arr), transf));
```
Comment (DebTalen) [inline src/mat_utils.cpp]: remove redundant parentheses

File: include/vse/mat_utils.hpp [hunk 11]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+
+#include <vec2.hpp>
```
Comment (DebTalen) [inline include/vse/mat_utils.hpp]: use public path <vse/*.hpp>

File: src/mat4.cpp [hunk 12]
```diff
@@ -0,0 +1,28 @@
+#include "mat4.hpp"
+#include <algorithm>
+
+mat4::mat4() {
```
Comment (DebTalen) [inline src/mat4.cpp]: maybe create an identity matrix by default and remove mat4(float) ctor?

File: include/vse/mat4.hpp [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+public:
+    mat4();
+    mat4(float x);
+    mat4(float arr[16]);
+    ~mat4() = default;
+
+    const float *const get_values() const { return values; }
+};
+
+mat4 mult(const mat4& lhs, const mat4& rhs);
```
Comment (DebTalen) [inline include/vse/mat4.hpp]: probably move to mat_utils?

**Comment by LevSmirnov1997** [comment]
@DebTalen 
branch updated with requeseted changes
[https://github.com/LevSmirnov1997/vse/tree/mat4](url)

File: src/mat4.cpp [hunk 14]
```diff
@@ -0,0 +1,10 @@
+#include "vse/mat4.hpp"
+#include <algorithm>
+
+mat4::mat4() :values{0.0f} {
```
Comment (DebTalen) [inline src/mat4.cpp]: move a member initializing list on the separate line

File: src/mat_utils.cpp [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    float t_arr[16];
+    std::fill_n(t_arr, 16, 0);
+    for(int i = 0; i < 4; ++i) {
+        for(int j = 0; j < 4; ++j) {
+            for(int k = 0; k < 4; ++k) {
+                t_arr[4*i+j] += lhs.get_values()[4*i+k] * rhs.get_values()[4*k+j];
+            }
+        }
+    }
+    return mat4(t_arr);
```
Comment (DebTalen) [inline src/mat_utils.cpp]: remove duplicated type declaration, use { t_arr };

**Pull request merged**